src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/18.diff.sql

```diff
@@ -0,0 +1,297 @@
+        SET @rightPartitionBoundary = DATEADD(hour, DATEDIFF(hour, 0, @currentDateTime) - @numberOfPartitions, 0);
+
+		-- Checks if a partition exists.
+		IF NOT EXISTS (SELECT 1 value FROM sys.partition_range_values AS prv
+						JOIN sys.partition_functions AS pf
+							ON pf.function_id = prv.function_id
+					WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+						and CONVERT(datetime2(7), prv.value, 126) = @rightPartitionBoundary) 
+		BEGIN
+			-- Creates new empty partition by creating new boundary value and specifying NEXT USED file group.
+			ALTER PARTITION SCHEME PartitionScheme_ResourceChangeData_Timestamp NEXT USED [Primary];
+			ALTER PARTITION FUNCTION PartitionFunction_ResourceChangeData_Timestamp() SPLIT RANGE(@rightPartitionBoundary); 
+		END;            
+            
+        SET @numberOfPartitions -= 1;
+    END;
+END;
+
+-- Creates a staging table 
+IF NOT EXISTS (SELECT 1 FROM sys.tables WHERE name = 'ResourceChangeDataStaging')
+BEGIN
+    -- Staging table that will be used for partition switch out. 
+    CREATE TABLE dbo.ResourceChangeDataStaging 
+    (
+        Id bigint IDENTITY(1,1) NOT NULL,
+        Timestamp datetime2(7) NOT NULL CONSTRAINT DF_ResourceChangeDataStaging_Timestamp DEFAULT sysutcdatetime(),
+        ResourceId varchar(64) NOT NULL,
+        ResourceTypeId smallint NOT NULL,
+        ResourceVersion int NOT NULL,
+        ResourceChangeTypeId tinyint NOT NULL,
+        CONSTRAINT PK_ResourceChangeDataStaging_TimestampId PRIMARY KEY (Timestamp, Id)
+    ) ON [PRIMARY]
+END;
+GO
+
+-- Creates check for a partition check.
+IF NOT EXISTS(SELECT 1 FROM sys.check_constraints WHERE name = 'chk_ResourceChangeDataStaging_partition') 
+BEGIN    
+    ALTER TABLE dbo.ResourceChangeDataStaging WITH CHECK 
+        ADD CONSTRAINT chk_ResourceChangeDataStaging_partition CHECK(Timestamp < CONVERT(DATETIME2(7), '9999-12-31 23:59:59.9999999'));
+
+    ALTER TABLE dbo.ResourceChangeDataStaging CHECK CONSTRAINT chk_ResourceChangeDataStaging_partition;
+END;
+GO
+    
+IF EXISTS (SELECT * FROM sys.indexes WHERE name = 'PK_ResourceChangeData')
+BEGIN
+    -- Drops index.
+    ALTER TABLE dbo.ResourceChangeData DROP CONSTRAINT PK_ResourceChangeData WITH (ONLINE = OFF);
+END;
+        
+IF NOT EXISTS (SELECT * FROM sys.indexes WHERE name = 'PK_ResourceChangeData_TimestampId')
+BEGIN    
+    -- Adds primary key clustered index.
+    ALTER TABLE dbo.ResourceChangeData ADD CONSTRAINT PK_ResourceChangeData_TimestampId 
+        PRIMARY KEY CLUSTERED(Timestamp ASC, Id ASC) ON PartitionScheme_ResourceChangeData_Timestamp(Timestamp)
+END;
+GO
+
+/*************************************************************
+    Purge partition feature for resource change data
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     ConfigurePartitionOnResourceChanges
+--
+-- DESCRIPTION
+--     Creates a staging table and initial partitions for the resource change data table.
```

Comment:
does this stored procedure actually create a staging table? or is the staging table created earlier via CREATE TABLE dbo.ResourceChangeDataStaging?

is there a way of validating that this stored procedure is working correctly?

EXEC dbo.ConfigurePartitionOnResourceChanges @numberOfFuturePartitionsToAdd = 48
select * from sys.partition_schemes

anything specific we should be looking for at this point in order to validate it is working correctly?